src/AsyncResourcePool.cs

```diff
@@ -152,29 +168,39 @@ private async void SetupPeriodicPurge()
                     timer.Dispose();
                 }
             }
-            
         }
 
         private ReusableResource<TResource> TryGetReusableResource()
         {
-            ReusableResource<TResource> reusableResource = null;
-            while (_availableResources.Count > 0 && reusableResource is null)
+            if (TryGetResource(out var resource))
+            {
+                return GetReusableResource(resource);
+            }
+
+            return null;
+        }
+
+        private bool TryGetResource(out TResource resource)
+        {
+            resource = default;
+            var result = false;
+            while (_availableResources.Count > 0 && !result)

```

Comment:
not really part of this feature, but I noticed that this is not a `ConcurrentQueue`... .this opens us up to concurrency bugs if two threads invoke `TryGetResource` simultaneously, right?

Reply:
The only methods that access non-thread safe fields are constrained by the `ActionBlock` (message handler). This is pretty much the actor pattern.

It would be possible to make a code change to this class that would break the thread-safety, but at the moment I'm not aware of any issues.

It might be worth splitting the class into a state container which can only be communicated with via the `ActionBlock`, and making the public methods (which may be called by multiple threads simultaneously) just pass messages to the inner container. This would make it clearer that the non-thread safe fields are constrained by the `ActionBlock`.

Also please keep unrelated changes in separate PRs -- if I wanted to make any changes to PR I'd end up re-writing your commit and it'd be kinda hard to collaborate on.

Reply:
I'll look again but just wanted to say this: please feel free to rewrite or drop ANY changes that I suggested on this PR - it was only as an easier way of "review". 
Will move the commit to a different PR.